docs: clarify Vault CA provider permissions needed (#15478)
hashicorp/consul: website/content/docs/connect/ca/vault.mdx

# Vault as a Service Mesh Certificate Authority

You can configure Consul to use [Vault](https://www.vaultproject.io/) as the certificate authority (CA), so that Vault can manage and sign certificates distributed to services in the mesh. 
The Vault CA provider uses the [Vault PKI secrets engine](https://www.vaultproject.io/docs/secrets/pki) to generate and sign certificates. 
This page describes how configure the Vault CA provider.

> **Tutorial:** Complete the [Vault as Consul Service Mesh Certification Authority](/consul/tutorials/vault-secure/vault-pki-consul-connect-ca) tutorial for hands-on guidance on how to configure Vault as the Consul service mesh certification authority.

## Requirements

- Refer to [Service Mesh Certificate Authority Overview](/docs/connect/ca) for important background information about how Consul manages certificates with configurable CA providers. 

- Use Vault 0.10.3 to latest 1.10.x.

> ~> **Note:** Do not use Vault v1.11.0+ as Consul's Connect CA provider — the intermediate CA becomes unable to issue the leaf nodes required by service mesh, and by Consul client agents if using auto-encrypt or auto-config and using TLS for agent communication. If you are already using Vault 1.11+ as a Connect CA, refer to this [Knowledge Base article](https://support.hashicorp.com/hc/en-us/articles/11308460105491) for more information about the underlying cause and recommended workaround.

Author: They are both Consul options/configurations. I also find the correction to be a bit confusing. I'd prefer to keep the additional wording.

Regarding the call-out text, same comment as above: I'll go with whatever you and @im2nguyen agree upon on generic vs specific.

> ~> **Note:** Do not use Vault v1.11.0+ as Consul's Connect CA provider — the intermediate CA becomes unable to issue the leaf nodes required by service mesh, and by Consul client agents if using auto-encrypt or auto-config and using TLS for agent communication. If you are already using Vault 1.11+ as a Connect CA, refer to this [Knowledge Base article](https://support.hashicorp.com/hc/en-us/articles/11308460105491) for more information about the underlying cause and recommended workaround.

Author: "Compatibility warning"

> ~> **Note:** Do not use Vault v1.11.0+ as Consul's Connect CA provider — the intermediate CA becomes unable to issue the leaf nodes required by service mesh, and by Consul client agents if using auto-encrypt or auto-config and using TLS for agent communication. If you are already using Vault 1.11+ as a Connect CA, refer to this [Knowledge Base article](https://support.hashicorp.com/hc/en-us/articles/11308460105491) for more information about the underlying cause and recommended workaround.

Author: This paragraph will also be touched again in about ~1 week, as we're actively working to remediate this incompatibility.

## Enable Vault as the CA

You can enable Vault as the CA by configuring Consul to use `"vault"` as the CA provider
and including the required provider configuration options.

> You can provide the CA configuration in the agent configuration file of server agents,

Author: I hesitated to write it this way to begin with because their are multiple types of files that can contain configuration ... "configuration entries" are different from "agent configuration".

That said, perhaps it's better overall to make this suggested change, even though there may be a possibility of misinterpretation.

</CodeTabs>

> ## Configuration Reference

Author: At some point, we should just make a pass over all the pages to change the capitalization.

For any _new_ headers I create, I'm using the style guide. But for content I'm not editing (or only shuffling on the page), I haven't been touching the existing header to make them compliant (because I'm focused on a particular change).

This probably creates a lot of back and forth for both reviewers and authors. If we update all the content at once, it should be easier for everyone going forward.

@include 'http_api_connect_ca_common_options.mdx'

### Root and Intermediate PKI Paths

The Vault CA provider uses two separately configured
[PKI secrets engines](https://www.vaultproject.io/docs/secrets/pki)
for managing Consul service mesh certificates.

The `RootPKIPath` is the PKI engine for the root certificate. Consul uses this root certificate to sign the intermediate certificate. Consul does not attempt to write or modify any data within the root PKI path.

The `IntermediatePKIPath` is the PKI engine used for storing the intermediate
signed with the root certificate. The intermediate signs all leaf
certificates, and Consul may periodically generate new intermediates for
automatic rotation. Therefore, Consul requires write access to this path.

> If neither path exists, then Consul attempts to mount and

Author: The original statement wasn't clear enough. The suggestion replacement isn't logically equivalent, though. I'll need to update.

for managing Consul service mesh certificates.

The `RootPKIPath` is the PKI engine for the root certificate. Consul uses this root certificate to sign the intermediate certificate. Consul does not attempt to write or modify any data within the root PKI path.

The `IntermediatePKIPath` is the PKI engine used for storing the intermediate
signed with the root certificate. The intermediate signs all leaf
certificates, and Consul may periodically generate new intermediates for
automatic rotation. Therefore, Consul requires write access to this path.

If neither path exists, then Consul attempts to mount and
initialize the PKI engine, which requires additional privileges of the Vault token in use.
If the paths already exist, Consul uses them as configured.

### Configure Vault ACL policies

You can define an access control list (ACL) policy that assigns Vault PKI management to Consul or Vault. If you want to manually create and tune the PKI secret engines used to store the root and intermediate certificates, define a policy that uses Vault-managed PKI paths. If you want to automate PKI management, define a policy that uses Consul-managed PKI paths.

> #### Define a policy for Vault-managed PKI paths

Author: This headings are often linked to. We need to create a link alias to avoid breaking existing links.

You can define a Vault ACL policy that allows Consul to use pre-existing PKI paths in Vault. The policy grants Consul read-only access to the PKI mount points and the root CA and grants full control of the intermediate or leaf CA for Consul service mesh clients.

In the following examples, the path configurations use placeholder values for `RootPKIPath` and `IntermediatePKIPath`. Replace them to match the path values in the CA provider configuration.

1. Add the following rules to you policy file so that Consul can read the PKI mounts already created in Vault:

>   <CodeBlockConfig filename="vault-managed-pki-policy.hcl">

Author: Note to self: go back through all of these and make sure root vs intermediate is used in all the right places.

Author: After checking, I fixed one mistake where I mixed up root and intermediate.

# to modify Consul's CA provider configuration as in (1) or (2).
# That's why this permission is currently commented out.
#
# path "<root_pki_path>/root/sign-self-issued" {
#  capabilities = [ "sudo", "update" ]
#}

# -----------------------------------------------------------------------------
# Needed for Consul to renew its Vault token (if renewable) whether provided
# directly in the CA provider config or using an auth method.

path "auth/token/renew-self" {
  capabilities = [ "update" ]
}
1. Only add the following rules to your policy file if you are changing the CA provider from Vault to a different CA provider (such as Consul's built-in provider) or changing the RootPKIPath used by the Vault CA provider:

path "auth/token/lookup-self" {
  capabilities = [ "read" ]
}
```
  <CodeBlockConfig filename="vault-managed-pki-policy.hcl">

</CodeBlockConfig>
  ```hcl
  path "<root_pki_path>/root/sign-self-issued" {
    capabilities = [ "sudo", "update" ]
  }
  ```
  </CodeBlockConfig>
  
  Note that because this is an extremely privileged endpoint, we recommend only temporarily granting this permission if you need to modify Consul's CA provider configuration.

> ### Consul Managed PKI Paths

Author: This headings are often linked to. We need to create a link alias to avoid breaking existing links.

the provided [Vault token](#token) must have the [required Vault privileges](#vault-acl-policies).
If the path does already exist, Consul uses the PKI secrets engine at that path as configured.

### Configure Vault ACL policies ((#vault-acl-policies))

The Vault CA provider requires a [Vault token](#token) with a specific set of Vault privileges
depending on whether you would prefer to control mount configuration from Vault or to delegate
that responsibility to Consul.

Use [Vault-managed PKI paths](#vault-managed-pki-paths) to obtain the following benefits:
- Enables use of a root CA external to Consul by instantiating the PKI secrets engine at
  [`RootPKIPath`](#rootpkipath) with an intermediate certificate
- Retain full control over PKI mount creation, and over `RootPKIPath` mount configuration

Otherwise, use [Consul-managed PKI paths](#consul-managed-pki-paths) to let Consul fully automate PKI management.

#### Define a policy for Vault-managed PKI paths ((#vault-managed-pki-paths))

To use Vault-managed PKI paths, you must first instantiate and configure PKI secrets engines at
`RootPKIPath` and `IntermediatePKIPath`.

> Next, define the following Vault ACL policy that allows Consul to use those pre-existing PKI mounts in Vault.
> The policy grants Consul limited use of the root PKI engine
> and full control of the intermediate PKI engine to issue service mesh certificates.

Reviewer: ```suggestion
Next, define a Vault ACL policy that allows Consul to use the pre-existing PKI mounts in Vault.
The policy grants Consul limited use of the root PKI engine
and full control of the intermediate PKI engine to issue service mesh certificates.
```
Aren't we repeating the same information in the steps below? I think we can delete this paragraph and the one above it. The next paragraph is a good introduction to the procedure for building a suitable policy.

Author: Yeah. I think I wrote this before I had the summary paragraph above that compares the two options.

    capabilities = [ "update" ]
  }
  ```

  </CodeBlockConfig>

1. Add the following rules to your Vault policy file so that Consul
   has read-only access to the root PKI engine, can automatically rotate
   intermediate CAs as needed, and has full use of the intermediate PKI engine:

  <CodeBlockConfig filename="vault-managed-pki-policy.hcl">

  ```hcl
  path "/<root_pki_path>/" {
    capabilities = [ "read" ]
  }
  
  path "/<root_pki_path>/root/sign-intermediate" {
    capabilities = [ "update" ]
  }

  path "/<intermediate_pki_path>/*" {
    capabilities = [ "create", "read", "update", "delete", "list" ]
  }
  ```
  </CodeBlockConfig>

1. Add the following rules to your Vault policy file so that Consul
   can renew its Vault token if the token is renewable.
   The rule enables the token to be renewed whether it is provided
   [directly](#token) in the CA provider configuration or presented in an [auth method](#authmethod).

  <CodeBlockConfig filename="vault-managed-pki-policy.hcl">

  ```hcl
  path "auth/token/renew-self" {
    capabilities = [ "update" ]
  }

  path "auth/token/lookup-self" {
    capabilities = [ "read" ]
  }
  ```
  
  </CodeBlockConfig>

> 1. Conditional and temporary: Only include the following rules in your Vault policy file
>    while you are changing the CA provider from Vault to a different CA provider (such as Consul's built-in provider)
>    or changing the `RootPKIPath` used by the Vault CA provider.
>    Those CA provider modifications trigger a root CA change that requires an
>    extremely privileged root cross-sign operation.
>    We recommend removing this elevated privilege immediately after the CA provider modification is finished.

Reviewer: ```suggestion
1. You should only include the following rules in your Vault policy file
   while changing the CA provider from Vault to a different CA provider, such as Consul's built-in provider,
   or changing the `RootPKIPath` used by the Vault CA provider.
   The CA provider modifications trigger a root CA change that requires a
   highly-privileged root cross-sign operation.
   Remove the elevated privilege immediately after the CA provider modification is finished.
```
I get why you wanted to label the step, but it adds too many words without as much benefit. As far as the recommendation to remove the elevated privileges, should that be a final step at the end of this procedure? It would make this step shorter and provide a dedicated action for readers following the process.

Author: The policy snippets in steps 1-3 must **always** be in place for Consul to work.

This policy snippet is ideally only in place for a very brief period of time while performing a specific operation.

These four "steps" aren't really a _procedure_ to follow. I don't think it makes sense to add a "step 5" for removing the elevated privileges.

Maybe the right solution is to move this temporarily privilege into its own section? It's shared between Vault and Consul managed paths, so I don't need to duplicate it in both sections... I'll try that out so we can take a look.

  <CodeBlockConfig filename="vault-managed-pki-policy.hcl">

  ```hcl
  path "<root_pki_path>/root/sign-self-issued" {
    capabilities = [ "sudo", "update" ]
  }
  ```
  </CodeBlockConfig>

#### Define a policy for Consul-managed PKI paths ((#consul-managed-pki-paths))

> To use Consul-managed PKI paths, ensure no PKI secrets engines are mounted at

Reviewer: Same comment as the previous procedure - we already describe what each rule in the policy does in the steps, so I think we should remove the first two paragraphs in this introduction.

  }

  path "/sys/mounts/<intermediate_pki_path>" {
    capabilities = [ "create", "read", "update", "delete", "list" ]
  }

  path "/sys/mounts/<intermediate_pki_path>/tune" {
    capabilities = [ "update" ]
  }
  ```
  </CodeBlockConfig>

1. Add the following rules to your Vault policy file so that Consul
   has full use of both PKI engines:

  <CodeBlockConfig filename="consul-managed-pki-policy.hcl">

  ```hcl
  path "/<root_pki_path>/*" {
    capabilities = [ "create", "read", "update", "delete", "list" ]
  }

  path "/<intermediate_pki_path>/*" {
    capabilities = [ "create", "read", "update", "delete", "list" ]
  }
  ```

  </CodeBlockConfig>

1. Add the following rules to your Vault policy file so that Consul
   can renew its Vault token if the token is renewable.
   The rule enables the token to be renewed whether it is provided
   [directly](#token) in the CA provider configuration or presented in an [auth method](#authmethod).

  <CodeBlockConfig filename="consul-managed-pki-policy.hcl">

  ```hcl
  path "auth/token/renew-self" {
    capabilities = [ "update" ]
  }

  path "auth/token/lookup-self" {
    capabilities = [ "read" ]
  }
  ```
  
  </CodeBlockConfig>

> 1. Conditional and temporary: Only include the following rules in your Vault policy file
>    while you are changing the CA provider from Vault to a different CA provider (such as Consul's built-in provider)
>    or changing the `RootPKIPath` used by the Vault CA provider.
>    Those CA provider modifications trigger a root CA change that requires an
>    extremely privileged root cross-sign operation.
>    We recommend removing this elevated privilege immediately after the CA provider modification is finished.

Reviewer: ```suggestion
1. You should only include the following rules in your Vault policy file
   while you are changing the CA provider from Vault to a different CA provider, such as Consul's built-in provider,
   or changing the `RootPKIPath` used by the Vault CA provider.
   The CA provider modifications trigger a root CA change that requires a
  highly-privileged root cross-sign operation.
   Remove the elevated privilege immediately after the CA provider modification is finished.
```
Same comment as above

   The rule enables the token to be renewed whether it is provided
   [directly](#token) in the CA provider configuration or presented in an [auth method](#authmethod).

  <CodeBlockConfig filename="consul-managed-pki-policy.hcl">

  ```hcl
  path "auth/token/renew-self" {
    capabilities = [ "update" ]
  }

  path "auth/token/lookup-self" {
    capabilities = [ "read" ]
  }
  ```
  
  </CodeBlockConfig>

#### Additional Vault ACL policies for sensitive operations

Additional Vault permissions are required while you perform the
following CA configuration changes:
- Changing the provider from Vault to a different provider, such as Consul's built-in provider
- Changing the `RootPKIPath`

Those configuration modifications trigger a root CA change that requires an
extremely privileged root cross-sign operation. 
For that operation to succeed, the CA's [Vault token](#token) or
[auth method](#authmethod) must contain the following rule:

<CodeBlockConfig filename="temporary-sensitive-operation-pki-policy.hcl">

```hcl
path "<root_pki_path>/root/sign-self-issued" {
capabilities = [ "sudo", "update" ]
}
```
</CodeBlockConfig>

> We recommend adding this elevated privilege just before the CA provider

Author: @rboyer : In practice... how would a Vault user actually accomplish this?

If I understand correctly, a Vault token's policies are immutable. That would mean you'd need to:
- Create a new Vault token with the elevated policy
- Apply the new Vault token to the Consul Connect CA configuration
- Perform the operation
- Reapply the previous Vault token with non-elevated permissions to the Consul Connect CA configuration

Alternatively, you could use a Vault auth method, and modify the policy associated with the auth method... But if I understand correctly, that wouldn't actually take effect unless Consul re-logs in using the auth method (after the modification) to get a new token (which contains the modified rules). And I don't know how the user would force Consul to do that... So maybe this is only workable when using a Vault token directly rather than an auth method, even though an auth method is generally preferable for enhanced security?

Consul is granted read-only access to the PKI mount points and the Root CA, but is
granted full control of the Intermediate or Leaf CA for Connect clients.

This example uses placeholder values for `RootPKIPath` and `IntermediatePKIPath`.
Please update those placeholders to match the path values in the CA provider configuration.

<CodeBlockConfig filename="vault-managed-pki-policy.hcl">

```hcl

> # -----------------------------------------------------------------------------

Author: Note to self: I'm still missing the guidance on what permissions are needed to change the provider away from the Vault CA provider.

That should probably be mentioned in the CLI/HTTP API docs as well... I'll address that in a future commit.

# Vault Managed PKI Mounts
# -----------------------------------------------------------------------------

> path "/sys/mounts" {

Author: As of [this PR](https://github.com/hashicorp/consul/pull/12655/files#diff-7bf30611a760296e2e6ffcd7fe955b0eb8317fe0a43c470b8b52d311af4a2515L359), I don't think we actually need read capabilities on `sys/mounts` anymore... There's no other call to `Sys().ListMounts()` as far as I can tell. (We should actually test this first.)

Can we remove that from the docs? Or at least mark the version upon which it is no longer necessary?

Author: It looks like this change was made in Consul 1.12.0: https://github.com/hashicorp/consul/blob/main/CHANGELOG.md#1120-april-20-2022

  capabilities = [ "read" ]
}

# -----------------------------------------------------------------------------
# Needed for Consul to renew its Vault token (if renewable) whether provided
# directly in the CA provider config or using an auth method.

> path "auth/token/renew-self" {

Author: Question for reviewer: do we think we should move the auth token renewal stuff to the bottom of the list of permissions? Maybe the most important stuff is the CA stuff itself.

(I think I agree with this. Will probably move to bottom unless told otherwise.)

Consul is granted read-only access to the PKI mount points and the Root CA, but is
granted full control of the Intermediate or Leaf CA for Connect clients.

> This example uses placeholder values for `RootPKIPath` and `IntermediatePKIPath`.
> Please update those placeholders to match the path values in the CA provider configuration.

Reviewer: ```suggestion
The `RootPKIPath` and `IntermediatePKIPath` paths have placeholder values. Replace them with real path values in your CA provider configuration.
```
In docs, fewer words outweighs politeness.

Author: Modified - check the latest wording

<CodeBlockConfig filename="vault-managed-pki-policy.hcl">

```hcl

> # -----------------------------------------------------------------------------

Reviewer: Curious about these in-file explanations. Do they mirror a template file included in some package?

Author: I didn't check what established practices we have for including comments in HCL file examples. JSON doesn't support comments, so it's not even possible in that context.

I included these comments so that, if someone copies and pastes this policy for their own usage, someone looking at it now or later can understand the purpose when it might not otherwise be obvious. In my experience, I've seen configuration file templates included with software include inline docs comments like this, including commented out things with an explanation of when it should be uncommented (like I've done here for the root rotation permission).

I'm open to removing the comments. I personally found the policy a lot more understandable with them than without them. It also helps with playing "spot the difference" between the Consul vs Vault managed mounts approaches.

Before proceeding, I'm curious what @boruszak and @im2nguyen think.

Reviewer: Here is more context about how I'm looking at the comment-based docs.

First, I do see the value. I have documented systems where the product ships with a commented up configuration file that the user duplicates and modifies to their implementation. It can be a useful way to present the information. 

For me, there are a few issues. As you pointed out, you can only provide the comments in HCL (and YAML), so we can't provide the examples for JSON configurations. We lose the rich text features, such as hyperlinks and text formatting, which might be useful in the explanations. Third, it's a deviation from our convention. Are users now expecting documentation within other example configuration files? 

I think the big thing, though, is that the information in this section is presented as a hybrid of conceptual and reference when we should present it as usage. There appears to be logical splits in the policy file where we could create "Define Vault managed PKI mount policies" section, for example. The section would include an explanation with links to the configuration reference items and an example. Or maybe it's a step-by-step process describing how to configure blocks of rules. When the information is presented as conceptual or reference, I think it reads as less actionable, so readers don't know if they should or have to do something.

I'll let Jeff and Tu chime in, though.

Author: If we went with the framing of "leave in comments where something may require thought/action", the only comment we'd want to leave in as the one about enabling root rotation, since best practices would suggest granting that permission for the minimum amount of time necessary when rotation is needed.

# Vault Managed PKI Mounts
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# Needed for Consul to read the PKI mounts already created in Vault

> path "/sys/mounts/<insert-RootPKIPath>" {

Reviewer 2: Since the `<>` already suggests "replace me" can we use the documented config field names:
```suggestion
path "/sys/mounts/<root_pki_path>" {
```

  capabilities = [ "read" ]
}

path "/sys/mounts/<insert-IntermediatePKIPath>" {
  capabilities = [ "read" ]
}

# Needed in Consul 1.11+ to update intermediate mount configuration
path "/sys/mounts/<insert-IntermediatePKIPath>/tune" {
  capabilities = [ "update" ]
}

# -----------------------------------------------------------------------------
# Needed for Consul to use the externally-managed, mounted PKI secrets engines

path "/<insert-RootPKIPath>/" {
  capabilities = [ "read" ]
}

path "/<insert-RootPKIPath>/root/sign-intermediate" {
  capabilities = [ "update" ]
}

path "/<insert-RootPKIPath>/*" {
  capabilities = [ "create", "read", "update", "delete", "list" ]
}

# Needed only when changing the CA provider from Vault with this RootPKIPath
# to either:
# 1. A different CA provider (such as Consul's built-in provider)
# 2. Changing the RootPKIPath used by the Vault CA provider
#
# SECURITY NOTE: Because this is an extremely privileged endpoint,
# We recommend only temporarily granting this permission if there's a need
# to modify Consul's CA provider configuration as in (1) or (2).
# That's why this permission is currently commented out.
#

> # path "<insert-RootPKIPath>/root/sign-self-issued" {

Author: @rboyer : This is my attempt to strike a balance between documenting this necessary permission while optimizing for security.

I also reference this from the "modify CA" CLI and HTTP API docs.

#  capabilities = [ "sudo", "update" ]
#}

# -----------------------------------------------------------------------------
# Needed for Consul to renew its Vault token (if renewable) whether provided
# directly in the CA provider config or using an auth method.

path "auth/token/renew-self" {
  capabilities = [ "update" ]
}

path "auth/token/lookup-self" {
  capabilities = [ "read" ]
}

# -----------------------------------------------------------------------------
# Deprecated

> # Only needed before Consul 1.12

Author: As of [this PR](https://github.com/hashicorp/consul/pull/12655/files#diff-7bf30611a760296e2e6ffcd7fe955b0eb8317fe0a43c470b8b52d311af4a2515L359), I don't think we actually need read capabilities on sys/mounts anymore... There's no other call to Sys().ListMounts() as far as I can tell. (We should actually test this first.)

It looks like this change was made in Consul 1.12.0: https://github.com/hashicorp/consul/blob/main/CHANGELOG.md#1120-april-20-2022

Reviewer 2: Can this be handled with versioned docs? I found the commented versioning guidelines for `/tune` confusing.

Reviewer 2: I support removing version references from config code blocks and calling out 1.12+ required 👍
